library/core/src/macros/mod.rs

```diff
@@ -14,10 +14,75 @@ macro_rules! panic {
         $crate::panicking::panic_str($msg)
     );
     ($fmt:expr, $($arg:tt)+) => (
-        $crate::panicking::panic_fmt($crate::format_args!($fmt, $($arg)+))
+        $crate::panicking::panic_fmt($crate::format_args!($fmt, $($arg)+), None)
     );
 }
 
+#[unstable(feature = "panic_internals", reason = "interal API", issue = "none")]
+#[macro_export]
+#[doc(hidden)]
+#[allow_internal_unstable(core_panic, panic_internals)]
+macro_rules! assert_binary_cmp {
+    ($macro_name:literal, $left:expr, $op:tt, $right:expr $(,)?) => ({
+        match (&$left, &$right) {
+            (left_val, right_val) => {
+                if !(*left_val $op *right_val) {
+                    // The reborrows below are intentional. Without them, the stack slot for the
+                    // borrow is initialized even before the values are compared, leading to a
+                    // noticeable slow down.
+                    let left_dbg: &dyn $crate::fmt::Debug = &&*left_val;
+                    let right_dbg: &dyn $crate::fmt::Debug = &&*right_val;
+                    $crate::panicking::panic_fmt(
+                        $crate::format_args!(r#"assertion failed: `(left {} right)`
+  left: `{:?}`,
+ right: `{:?}`"#, stringify!($op), left_dbg, right_dbg),
+                        $crate::option::Option::Some($crate::panic::ExtraInfo::AssertInfo(&$crate::panic::assert_info::AssertInfo {
+                              assertion: $crate::panic::assert_info::Assertion::Binary($crate::panic::assert_info::BinaryAssertion {
+                                  op: stringify!($op),
+                                  left_expr: stringify!($left),
+                                  right_expr: stringify!($right),
+                                  left_val: left_dbg,
+                                  right_val: right_dbg,
+                              }),
+                              macro_name: $macro_name,
+                              message: $crate::option::Option::None,
+                        })),
+                    );
+                }
+            }
+        }
+    });
+    ($macro_name:literal, $left:expr, $op:tt, $right:expr, $($msg:tt)*) => ({
+        match (&$left, &$right, $crate::format_args!($($msg)*)) {
```

Comment:
`format_args!` is now called before comparing the values. Could this be a problem?

Reply:
Hmm, yeah, it could be. I'm not sure how much work runtime over `format_args!` has if you don't actually format it, but it certainly seems wrong.